sirfilesystem.c

```diff
@@ -396,8 +422,158 @@ bool _sir_getrelbasepath(const char* restrict path, bool* restrict relative,
     return true;
 }
 
+#if defined(_AIX)
+# define SIR_MAXSLPATH SIR_MAXPATH + 16
+int _sir_aixself(char* buffer, size_t* size) {
+    ssize_t res;
+    char cwd[SIR_MAXPATH], cwdl[SIR_MAXPATH];
+    char symlink[SIR_MAXSLPATH], temp_buffer[SIR_MAXPATH];
+    char pp[64];
+    struct psinfo ps;
+    int fd;
+    char** argv;
+    char* tokptr;
+
+    if ((buffer == NULL) || (size == NULL))
+        return -1;
+
+    snprintf(pp, sizeof(pp), "/proc/%llu/psinfo", (unsigned long long)_sir_getpid());
+
+    fd = open(pp, O_RDONLY);
+    if (fd < 0)
+        return -1;
+
+    res = read(fd, &ps, sizeof(ps));
+    close(fd);
+    if (res < 0)
+        return -1;
+
+    if (ps.pr_argv == 0)
+        return -1;
+
+    argv = (char**)*((char***)(intptr_t)ps.pr_argv);
+
+    if ((argv == NULL) || (argv[0] == NULL))
+        return -1;
+
+    if (argv[0][0] == '/') {
+        snprintf(symlink, SIR_MAXPATH, "%s", argv[0]);
+
+        /* Flawfinder: ignore */
+        res = readlink(symlink, temp_buffer, SIR_MAXPATH);
+        if (res < 0)
+            strlcpy(buffer, symlink, SIR_MAXPATH);
+        else
+            snprintf(buffer, *size - 1, "%s/%s", (char*)dirname(symlink), temp_buffer);
+
+        *size = strlen(buffer);
+        return 0;
+    } else if (argv[0][0] == '.') {
+        char* relative = strchr(argv[0], '/');
+        if (relative == NULL)
+            return -1;
+
+        snprintf(cwd, SIR_MAXPATH, "/proc/%llu/cwd", (unsigned long long)_sir_getpid());
+
+        /* Flawfinder: ignore */
+        res = readlink(cwd, cwdl, sizeof(cwdl) - 1);
+        if (res < 0)
+            return -1;
+
+        snprintf(symlink, SIR_MAXPATH, "%s%s", cwdl, relative + 1);
+
+        /* Flawfinder: ignore */
+        res = readlink(symlink, temp_buffer, SIR_MAXPATH);
+        if (res < 0)
+            strlcpy(buffer, symlink, SIR_MAXPATH);
+        else
+            snprintf(buffer, *size - 1, "%s/%s", (char*)dirname(symlink), temp_buffer);
+
+        *size = strlen(buffer);
+        return 0;
+    } else if (strchr(argv[0], '/') != NULL) {
+        snprintf(cwd, SIR_MAXPATH, "/proc/%llu/cwd", (unsigned long long)_sir_getpid());
+
+        /* Flawfinder: ignore */
+        res = readlink(cwd, cwdl, sizeof(cwdl) - 1);
+        if (res < 0)
+            return -1;
+
+        snprintf(symlink, SIR_MAXPATH, "%s%s", cwdl, argv[0]);
+
+        /* Flawfinder: ignore */
+        res = readlink(symlink, temp_buffer, SIR_MAXPATH);
+        if (res < 0)
+            strlcpy(buffer, symlink, SIR_MAXPATH);
+        else
+            snprintf(buffer, *size - 1, "%s/%s", (char*)dirname(symlink), temp_buffer);
+
+        *size = strlen(buffer);
+        return 0;
+    } else {
+        char clonedpath[16384];
```

Comment:
confusing magic number

Reply:
It's from AIX example code, so I don't want to change it :)